### PR TITLE
Build, CI & Vendoring Improvements

### PR DESCRIPTION
**Pull request by lrishi**
This PR does following:

1. Add support for multiarch builder images
2. Vendor zlib and remove all dynamic links
3. Prepare for toolchain-less build
4. Workflow updates

### CODE REVIEW

**Review by mmat11 — APPROVED** [review]

**Comment by stanek-michal** [comment]
Why did we need to vendor in zlib? I think I saw this mentioned in some thread but I can't find it now. Is it so that we don't have to install zlib in the build containers?

Out of curiosity, are we planning to use Zig for eBPF?

**Comment by lrishi** [comment]
> Why did we need to vendor in zlib? I think I saw this mentioned in some thread but I can't find it now. Is it so that we don't have to install zlib in the build containers?
> 
> Out of curiosity, are we planning to use Zig for eBPF?

To do away with this line:
`set_property(TARGET libbpf PROPERTY INTERFACE_LINK_LIBRARIES "libelf;-lz")`
libbpf uses libz, we were linking with the os version until now, which was a big headache if copying binaries into containers

Yes, zig is really simple to use, and brings in the latest clang/llvm toolchain.

File: Makefile [hunk 1]
```diff
@@ -1,79 +1,113 @@
...
-	cmake ${EXTRA_CMAKE_FLAGS} ${CMAKE_FLAGS}
-	make VERBOSE=1 -C${BUILD_DIR} -j$(shell nproc)
+	cmake ${EXTRA_CMAKE_FLAGS} ${CMAKE_FLAGS} -B${BUILD_DIR} -S${PWD}
+	cmake --build ${BUILD_DIR} --parallel $(shell nproc)
+	@echo -e "\n++ Build Successful at `date` ++\n"
+else
+ifdef BUILD_CONTAINER_IMAGE
+	make container
+endif
+	${CONTAINER_RUN_CMD} \
```
Comment (nicholasberlin) [inline Makefile:55]: Is this line ending `\` intentional?

File: Makefile [hunk 2]
```diff
@@ -1,79 +1,113 @@
...
+	make build DEBUG=${DEBUG} ARCH=${ARCH} EXTRA_CMAKE_FLAGS=${EXTRA_CMAKE_FLAGS}
+	make fix-permissions
+endif
 
 container:
-	docker build -t ${DOCKER_LOCAL_TAG} -f docker/Dockerfile.builder .
+	${CONTAINER_ENGINE} buildx build --progress plain --platform=linux/${ARCH} -t ${CONTAINER_LOCAL_TAG} -f docker/Dockerfile.builder .
```
Comment (nicholasberlin) [inline Makefile:61]: Is `buildx` supposed to be there? (I see it in a few other places as well, so maybe it is intentional)

File: GPL/HostIsolation/TcFilter/CMakeLists.txt [hunk 3]
```diff
@@ -24,7 +24,7 @@ set(TCFILTER_CFLAGS
 
 set(TCFILTER_BPF_OUTPUT ${HOSTISOLATION_EBPF_OBJECT_DIR}/TcFilter.bpf.o)
 add_custom_command(OUTPUT ${TCFILTER_BPF_OUTPUT}
-                  COMMAND ${EBPF_EXTERNAL_ENV_FLAGS} ${CLANG} ${TCFILTER_CFLAGS} -emit-llvm -c ${CMAKE_CURRENT_SOURCE_DIR}/TcFilter.bpf.c -o - | ${LLC} -march=bpf -mcpu=v2 -filetype=obj -o  ${TCFILTER_BPF_OUTPUT}
```
Comment (stanek-michal) [inline GPL/HostIsolation/TcFilter/CMakeLists.txt:27]: Why can't we leave the `mcpu` flag? I remember passing `mcpu=v2` was necessary to get TcFilter to work on older kernels.

File: docker/Dockerfile.builder [hunk 4]
```diff
@@ -1,45 +1,132 @@
...
+# Stage 1: Build static bpftool
+# This step is required specifically to maintain support for
+# centos7. bpftool is available on newer distros via package
+# manager, but elastic/ebpf is built on centos7 to workaround
+# glibc's unsupported backwards-compatibility. Here, we build
+# a static bpftool on a newer ubuntu and copy it to centos7
+# builder image to generate skeleton headers.
+
+FROM ubuntu:jammy as bpftool-builder
+ARG SCRATCH
```
Comment (nicholasberlin) [inline docker/Dockerfile.builder:12]: Does this line play nicely with line 1?

File: GPL/HostIsolation/KprobeConnectHook/CMakeLists.txt [hunk 5]
```diff
@@ -7,17 +7,18 @@
 
 set(KPROBECONNECTHOOK_CFLAGS
                         -g -O2
-                        -target bpf
```
Comment (stanek-michal) [inline GPL/HostIsolation/KprobeConnectHook/CMakeLists.txt:10]: Don't we need this flag? When I try to remove it and build with Endpoint I'm getting this:
```
error: unable to create target: 'No available targets are compatible with triple "x86_64-unknown-linux-gnu"'

**Review by nicholasberlin — APPROVED** [review]
Seems fine to me if it works.

File: GPL/HostIsolation/TcFilter/CMakeLists.txt (hunk 3 above)
Comment (lrishi) [reply GPL/HostIsolation/TcFilter/CMakeLists.txt:27]: Added it back 👍🏽

File: GPL/HostIsolation/KprobeConnectHook/CMakeLists.txt (hunk 5 above)
Comment (lrishi) [reply GPL/HostIsolation/KprobeConnectHook/CMakeLists.txt:10]: oh this wont build with endpoint right away, I have a separate diff that i'll PR once this is in.
This flag is now a part of COMPILER_FLAGS, there are minor variations in the way zig and clang accept these flags.

File: Makefile (hunk 1 above)
Comment (lrishi) [reply Makefile:55]: yup

File: Makefile (hunk 2 above)
Comment (lrishi) [reply Makefile:61]: yup

File: docker/Dockerfile.builder (hunk 4 above)
Comment (lrishi) [reply docker/Dockerfile.builder:12]: yup, just have to redefine the name to use it after a 'FROM' instruction.

**Review by stanek-michal — APPROVED** [review]
Looks good, thanks for the changes!

**Pull request merged**